Import any type of XBlock into Blockstore [SE-279] [LX-73]

bradenmacdonald opened the pull request:
This improves the XBlock/XModule import process.

Before, only a few block types (course, chapter, sequential, vertical, problem, video, html) were supported. Now any block type works, whether it's an XBlock or an XModule.

Changes:
* Use the newer `add_xml_to_node` XBlock API (as opposed to the old `definition_to_xml`) XModule API), since it works with all block types.
* Pretty-prints the OLX XML
* Use edx-platform's safe/defused libxml instead of python's built-in `xml` module
* Removed special handling to export video transcripts - now edx-val does it for us
* Changes the test suite to actually call edX APIs instead of mocking everything. This creates a much more useful and realistic test, but the tests are much slower to run and must be run in the Studio devstack environment.
* Changes the organization of files in the resulting bundle
* Ensure there is always a `url_name` for each block XML node in the OLX.

Side effects: The newer `add_xml_to_node` API that I'm using here always exports a block _and all its children_, so importing a unit to blockstore now necessarily creates a single `.olx` file instead of one main `.olx` file and one per child block. This may be better, because we want blockstore to edit at the granularity of author intent.

Open questions:

1. The `definition_to_xml` API includes a `resource_fs` parameter that allows XModules to export data files along with their XML (e.g. subtitle files, video files, etc.). AFAIK, the XBlock API has no equivalent mechanism. How should pure XBlocks deal with resource files in the future? (My suggestion is to create a new FileField field type which can reference a file or a set of files, which will have numerous benefits by making the runtime and blockstore able to be aware of every static file needed by each block.)  (Currently, edx-platform works around this by setting an `export_fs` attribute on the XBlock runtime, so the XModule implementation of `add_xml_to_node` passes that in to each descriptor's `definition_to_xml` method.)

2. ~~How do we (how does an XBlock runtime backed by Blockstore) reference a block contained within a vertical's OLX file within a bundle? e.g. if I create `<vertical url_name="v"><drag-and-drop-v2 url_name="bob"/></vertical>` within Studio and import it into Blockstore, the `bundle.json` will only hold a reference to `/vertical/v.olx` and there is no easy way to find the drag and drop XBlock within the bundle without iterating over all verticals in the bundle. Presumably then, we would not consider a bundle name + block type + url_name sufficient to identify an XBlock and would instead need to use a new form of identifier like `bundle:/vertical/v.olx:drag-and-drop-v2:bob`~~ (let's see if we actually need this...)

3. ~~Do we allow importing leaf XBlocks like a bare `html` block (not wrapped in a vertical) into a bundle? If so, how is that handled? (Should it appear in the `bundle.json` `components` list? Should it be auto-wrapped in a `<vertical>`? How does one refer to it? If one later imports a vertical that contains that block, should blockstore dereference it so that the vertical's OLX points to the `html` block's own `.olx` file?)~~

4. ~~Does this break anything? I haven't tested it much. Not sure if static assets or video files still work.~~

## Test instructions

Install according to the README.

Open the Studio shell (`make studio-shell`) then run the test "suite" with:

```
python -m pytest --ds=cms.envs.test --no-cov --nomigrations /edx/src/openedx-blockstore-relay/openedx_blockstore_relay/tests/
```

You can also test importing a unit from the studio shell using the instructions in the README.

bradenmacdonald committed:
WIP

bradenmacdonald commented:
@symbolist @ormsbee @pomegranited I would welcome your input on this PR and the list of questions I wrote up, while I play around with connecting an XBlock runtime to Blockstore.

ormsbee commented:
1. I'm a bit leery of adding `FileField` because I think that XBlock field data has been a bit abused, and we've run into weird edge cases where certain things aren't possible with things like the pyfilesystem field type (e.g. field data overrides). In general, I don't like how XBlock tries to hide things from the developer via a field interface (maybe it's in this database, maybe it's in that database, maybe it's in S3) because I feel that it makes the overall system more complex -- even if it makes the interface simpler. All that being said, XBlocks need to have some way to talk about these things and fields are the most idiomatic thing given the state of the design today. So count me as a "I wish there was a better way forward, but there probably isn't" +1 on this.
2. Do we need this addressability for student state storage, links, or something else...? I totally believe that we need it, but I'm not sure what all the use cases are.
3. I think we should allow it, without auto-wrapping.

symbolist commented:
1. Hmm. Actually don't really have sufficient details of the usages in my head for this.
2. If we want to do this, the better way would be to use something standard like XPath (together with the xinclude idea). That way, we can use standard libraries and even Blockstore will be able to understand nesting of entities without needing a runtime. Though one potential issue I see is with identifiers; If every OLX file is a separate file, we can use the path for deriving the identifier but how would it work if XBlocks are nested in one file? And if children Xblocks get added or removed from that file?
3. I think Product will have opinions here because the decisions will impact how the UI works and what is possible. Though I would expect authors to want to be able to export leaf nodes, and decide if only the vertical or also some of its children should be separately reference-able (that is the intention of the `components` field; i.e. give authors the flexibility to control this).

When the initial work was being done, the tricky part turned out to be how to figure out which assets belong to a specific XBlock. In course exports, all assets are exported into a directory and so that works. How would this work when exporting individual blocks of arbitrary type? I see your point 1, but that would require changing how XBlocks store data so I don't think we plan to use it for existing content?

pomegranited commented:
@bradenmacdonald 

1. I think making an XBlock's file resources more explicit would be much better than the implicit way we do it now -- e.g. the video xmodule's files are all over the place.
   But we don't want XBlock authors using it as a file store for student data, so wouldn't want it to be scopable as `user_state`, so I don't think a `FileField` is a good idea either..
   Maybe add the capability via an XBlock mixin instead?  I wasn't able to find a better library than `pyfilesystem` though..
   > certain things aren't possible with things like the pyfilesystem field type (e.g. field data overrides)

   @ormsbee  Can you elaborate?
2. @symbolist 👍 for using standard addresses like XPath and xinclude.
   > If every OLX file is a separate file, we can use the path for deriving the identifier but how would it work if XBlocks are nested in one file? And if children Xblocks get added or removed from that file?

   I thought we'd have the problem of managing the addition/removal of child XBlocks regardless of whether we're storing in a single OLX file or not?
   Maybe the bundle can expose its tree's paths in its `bundle.json`, then no one else has to traverse the full OLX to find things.  And then it's quick to find out what has been added/removed between versions.
3. Yep, I think course authors will want to be able to import XBlocks that aren't wrapped in verticals.
   > Should it appear in the bundle.json components list? 

   By default, yes.  I think authors should be able to edit this `components` list, as it is the public/importable interface to the bundle.
   > Should it be auto-wrapped in a <vertical>? 

   Does it need to be?  If not, then why wrap it?
   > If one later imports a vertical that contains that block, should blockstore dereference it so that the vertical's OLX points to the html block's own .olx file?

   If someone re-imports a block that matches an existing ID, then that existing block should be overwritten.  In the GUI, we can possible let them know this is about to happen so they won't lose local changes?
4. I imported the edX Demo course into a blockstore collection using this branch, and it seemed to work fine. 
  But I still got xml files created for each child (and .html files for the HTML XModules) added to my bundle?  There's a `course.olx` file, but it just contains references to the child `<chapter>` and `<wiki>` elements.
    ```bash
    root@5e593b372241:/edx/app/edxapp/edx-platform# ./manage.py cms transfer_to_blockstore --settings=devstack_docker --block-key=block-v1:edX+DemoX+Demo_Course+type@course+block@course --collection-uuid=<collection-uuid>
    ```
    *Note*: For this to run without error in my docker devstack, I have to set `common.lib.xmodule.xmodule.video_module.edxval_api = None`, else I get errors:
   ```
   IOError: [Errno 2] No such file or directory: u'/edx/var/edxapp/media/video-transcripts/34121ac6acbd4cc0a14a5ce6ff873194.sjson'
   ```

bradenmacdonald committed:
Address review comments

bradenmacdonald committed:
Update tests, now tests need actual edxapp API

bradenmacdonald committed:
Test uploaded bundle file contents

bradenmacdonald committed:
Organize bundle files better

bradenmacdonald commented:
@symbolist @pomegranited @ormsbee 

Thanks for the responses everyone. I have updated this PR:

* Fixed the issues Jill reported ("still got xml files created for each child", and edx-val related errors) - although there turns out to be no clean way to prevent the HTML XModule from saving out to `.html` files, so that still happens (Since HTML is not a subset of XML, that's probably a good thing? Otherwise it would need to be in a CDATA section...)
* Modified the test suite to actually talk to modulestore and export a real course. This means that you must run the tests from within a Studio devstack and it does make them slow to run, but I believe it makes the tests dramatically more helpful, while arguably being cleaner than the mocked version. See the code to see what I mean.
* Expanded the tests to check the exported OLX content
* **Changed the layout of the bundle contents**: Now, importing a `vertical` with the ID of `foo` puts the OLX for that vertical and all the random associated `.html` and `.srt` files inside a folder called `vertical-foo`. Any _public_ static assets referenced in e.g. html modules still go into a top-level `/static/` folder to reflect the fact that they are public, not private, and may be used by multiple units.

@symbolist Can you please review?

bradenmacdonald committed:
Fix bug when root imported block is a pure XBlock

ormsbee commented:
>> certain things aren't possible with things like the pyfilesystem field type (e.g. field data overrides)
>
> @ormsbee Can you elaborate?

Sorry I lost track of this one. It's one of the more obscure corners of XBlock, so please forgive me if I'm saying stuff you already know.

I meant the example of the little used [`Filesystem`](https://github.com/edx/XBlock/blob/9c6021281f07b9f2d817a795d90ea9d1f3530987/xblock/reference/plugins.py#L109) field type. It's a Field that represents a `pyfilesystem`. It's sometimes used with a `user_state_summary` scope in order to give all the users of a block a shared virtual filesystem. That filesystem can then be used for collaboration and even messaging, with directories by group or individual user and message creation as files.

It's a powerful tool that builds on the `fs` service. But in retrospect, it's not clear to me that having it as a `Field` was a particularly useful abstraction. XBlocks could have still have implemented that sort of functionality by using the runtime service directly, and `Filesystem` as a `Field` is a little peculiar. The `__set__` and `__delete__` methods had to be overridden to raise `NotImplementedError` as they're kind of nonsensical when talking about a whole file system. It's not clear what it means to override such a thing with field overrides in the CCX model. We took a new concept and shoehorned it into the shape of a `Field` because that's one of the basic components of XBlocks, but we made something that only really acts like a field half the time. It's a Field with a giant asterisk at the end. Cale and I have definitely had conversations where we've said, "Yeah, that sounds like it'll work... oh, except the Filesystem field..."

I feel like it's one of those "when all you have is a hammer, everything looks like a nail" situations. Fields are a giant hammer. Runtime services are too, frankly. I agree that it's desirable to have some way to make the XBlock aware of the static assets that it's using. Fields seem like a plausible way to do that. I'm just wondering if there's a better tool for it.

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -37,6 +39,40 @@ def transfer_to_blockstore(block_key, bundle_uuid=None, collection_uuid=None):
+
+    old_export_fs = block.runtime.export_fs
+    block.runtime.export_fs = fs
+    if hasattr(block, 'export_to_file'):
```
Are there any cases for this besides the one in `CustomTagDescriptor`?

bradenmacdonald replied:
This affects most XBlocks and XModules in edx-platform, so it has many use cases for us.

But no, AFAIK, `CustomTagDescriptor` and this code here are the only places where this is `False`.

bradenmacdonald commented:
BTW at this point I think the main open question on this PR, related to other recent discussions, is: what file path should be used for the OLX files that are imported? Currently this PR changes the path to `/{block_type}-{block_url_name}/{block_url_name}.olx`, but @symbolist has also suggested that `/{block_url_name}.olx` might be better, and I do like that idea.

bradenmacdonald committed:
Fix indentation

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -103,42 +139,27 @@ def block_key(self):
-                child_ref = ElementTree.Element(child.location.block_type)
-                child_ref.set('url_name', child.url_name)
-                olx_node.append(child_ref)
+        olx_node = Element("unknown_root", nsmap=XML_NAMESPACES)
```
What is the significance of "unknown_root"?

bradenmacdonald replied:
This code is copied from the `XBlock` code repo. The name of the node doesn't matter since the XBlock will rename the node when it adds its data to it.

symbolist replied:
Ah okay! Maybe put in a note for future readers of this code?

symbolist commented on openedx_blockstore_relay/tests/test_transfer_data.py:
```diff
@@ -9,125 +9,108 @@
+        content = StaticContent(asset_key, "Fake asset", "application/text", "test".encode('utf8'))
+        contentstore().save(content)
+        # And the video data + transcript must also be stored in edx-val for the video export to work:
+        edxval_api.create_video(VIDEO_B_VAL_DATA)
```
If we export a course from edx.org and import it into stage.labxchange.com edx instance, which will not have VEDA configured, will this still work?

bradenmacdonald replied:
I was under the impression that edx-val was enabled by default from Hawthorn onward?

There is no feature flag for edx-val that I can see, it's simply [enabled if installed](https://github.com/edx/edx-platform/blob/0cfda2d09782ba0ddb3ee2632e5101ad6491f9f1/common/lib/xmodule/xmodule/video_module/video_module.py#L84-L87), and [it's installed by default](https://github.com/edx/edx-platform/blob/66b1c07b7195f9ffd9b9e211384465bce2294d3b/requirements/edx/base.in#L86). We may not use VEDA, but we will be using VAL, and I think VAL is all that's needed here.

symbolist replied:
Oh okay, that would be great.

symbolist commented on openedx_blockstore_relay/tests/README.rst:
```diff
@@ -0,0 +1,6 @@
+
+Run these tests from the devstack CMS shell (``make studio-shell``) using::
+
+    python -m pytest --ds=cms.envs.test /edx/src/openedx-blockstore-relay/openedx_blockstore_relay/tests/
```
We really need it on CI for a number of repos. Do you think it is feasible?

bradenmacdonald replied:
It would be feasible if edX published docker images of `edx-lms` and `edx-studio` that (combined with standard `mysql`, `memcached`, etc. images) run easily. Then it would simply be a matter of configuring CI to run the `edx-studio` image, install this app into that image's virtualenv, then run the tests.

I believe the main work involved is getting the tooling in place to auto-publish working `edx-studio` container to Docker Hub. They only currently offer a devstack `edxapp` image, which is not self-contained.

symbolist replied:
I see. Though given the emphasis on having code in separate packages + plugin based architecture, I am surprised edX hasn't made this straightforward already.

symbolist commented:
@bradenmacdonald Okay, this higher level API sounds a lot cleaner! I was worried we would have to special case a lot of stuff.

> what file path should be used for the OLX files that are imported?

Hmm. A video block may have tens of transcripts. Okay, if we never expect a human to edit the bundle contents, it should be fine to put everything at the root. But I can see a case for per XBlock directory too. Another consideration is that static files likes images which are included in html content may be used by more than one block so should they go into a common dir?

And we can always come back to this after the authoring runtimes have progressed enough that we have a better idea?

> It's a Field with a giant asterisk at the end. Cale and I have definitely had conversations where we've said, "Yeah, that sounds like it'll work... oh, except the Filesystem field..."

@ormsbee Maybe this makes a case for keeping the new BlockstoreRuntime completely separate from the old runtimes and use the opportunity to deprecate such stuff?

bradenmacdonald commented:
@symbolist  Yeah, we could also put the .OLX file at the root, and any files specific to that OLX file could be in a subdir, so the layout could be:

```
/unit1.olx
/unit1/video_subtitle1.srt
/unit1/video_subtitle2.srt
/unit2.olx
/problem.olx
/static/images_used_by_one_or_more_blocks.jpg
```

symbolist commented:
@bradenmacdonald Okay, that looks nice! Most blocks do not have files (besides the static ones) so the few who do can have directories if needed. I like it. 👍

bradenmacdonald committed:
Add url_name to pure XBlocks in the OLX

bradenmacdonald committed:
PEP8 fixes

bradenmacdonald committed:
Simplify OLX file paths

bradenmacdonald commented:
@symbolist OK, updated to use that convention. Ready for another review.

symbolist commented on openedx_blockstore_relay/adapters.py:
```diff
@@ -0,0 +1,75 @@
+        For exporting, set data on `node` from ourselves.
+        """
+        super(UrlNameMixin, self).add_xml_to_node(node)
+        if not node.get('url_name'):
```
We don't have to make any change now, but just wondering if we are planning to clean this attribute name to something like `id`?

bradenmacdonald replied:
I do think that something like `id` would make more sense. I think we had discussed changing that too, but for now we have no specific plans.

symbolist reviewed (approved):
Do the tests take a long time to run for you too? I see around 90 seconds for a run. But besides that looks good to go! 👍

bradenmacdonald committed:
Update test instructions to be much faster

bradenmacdonald commented:
@symbolist Thanks! 

> Do the tests take a long time to run for you too? I see around 90 seconds for a run.

Yeah, I mentioned in the description that the tests take a long time to run. But, since you raised the issue, I investigated it a bit more. I found that of those 90 seconds, 60s is spent applying database migrations and 30s is spent analyzing coverage. So the test can run in only a few seconds with a modified command line:

```
python -m pytest --ds=cms.envs.test --no-cov --nomigrations /edx/src/openedx-blockstore-relay/openedx_blockstore_relay/tests/
```

I updated the test README file accordingly.
